tensorboard/util/grpc_util.py

```diff
@@ -51,7 +53,56 @@
 _VERSION_METADATA_KEY = "tensorboard-version"
 
 
-def async_call_with_retries(api_method, request, done_callback, clock=None):
+class AsyncCallFuture:
+    """Encapsulates the future value of a retriable async gRPC request.
+
+    Abstracts over the set of futures returned by a set of gRPC calls
+    comprising a single logical gRPC request with retries.  Communicates
+    to the caller the result or exception resulting from the request.
+
+    Args:
+      completion_event: The constructor should provide a thredding.Event which
+        will be used to communicate when the set of gRPC requests is complete.
+    """
+
+    def __init__(self, completion_event):
+        self._active_grpc_future = None
+        self._active_grpc_future_lock = threading.Lock()
+        self._completion_event = completion_event
+
+    def _set_active_future(self, grpc_future):
+        if grpc_future is None:
+            raise RuntimeError(
+                "_set_active_future invoked with grpc_future=None"
+            )
+        # TODO() maybe add validation logic here that future is not None.
+        with self._active_grpc_future_lock:
+            self._active_grpc_future = grpc_future
+
+    def result(self, timeout):
+        """Analogous to `grpc.Future.result`. Returns the value or exception.
+
+        This method will wait until the full set of gRPC requests is complete
+        and then act as `grpc.Future.result` for the single gRPC invocation
+        corresponding to the first successful call or final failure, as
+        appropriate.
+
+        Args:
+          timeout: How long to wait in seconds before giving up and raising
+
+        Returns:
+          The result of the future corresponding to the single gRPC
+          corresponding to the successful call or final failure, as
```

Comment:
In a failure case this will raise an exception instead.  So probably `Returns:` should cover the success case and we should have `Raises:` for the failure case.